homeassistant/components/devolo_home_network/button.py

```diff
@@ -0,0 +1,132 @@
+    IDENTIFY: DevoloButtonEntityDescription(
+        key=IDENTIFY,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        icon="mdi:led-on",
+        name="Identify device with a blinking LED",
+        press_func=lambda device: device.plcnet.async_identify_device_start(),  # type: ignore[union-attr]
+    ),
+    PAIRING: DevoloButtonEntityDescription(
+        key=PAIRING,
+        icon="mdi:plus-network-outline",
+        name="Start PLC pairing",
+        press_func=lambda device: device.plcnet.async_pair_device(),  # type: ignore[union-attr]
+    ),
+    RESTART: DevoloButtonEntityDescription(
+        key=RESTART,
+        device_class=ButtonDeviceClass.RESTART,
+        entity_category=EntityCategory.CONFIG,
+        name="Restart device",
+        press_func=lambda device: device.device.async_restart(),  # type: ignore[union-attr]
+    ),
+    START_WPS: DevoloButtonEntityDescription(
+        key=START_WPS,
+        icon="mdi:wifi-plus",
+        name="Start WPS",
+        press_func=lambda device: device.device.async_start_wps(),  # type: ignore[union-attr]
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Get all devices and buttons and setup them via config entry."""
+    device: Device = hass.data[DOMAIN][entry.entry_id]["device"]
+
+    entities: list[DevoloButtonEntity] = []
+    if device.plcnet:
+        entities.append(
+            DevoloButtonEntity(
+                entry,
+                BUTTON_TYPES[IDENTIFY],
+                device,
+            )
+        )
+        entities.append(
+            DevoloButtonEntity(
+                entry,
+                BUTTON_TYPES[PAIRING],
+                device,
+            )
+        )
+    if device.device and "restart" in device.device.features:
+        entities.append(
+            DevoloButtonEntity(
+                entry,
+                BUTTON_TYPES[RESTART],
+                device,
+            )
+        )
+    if device.device and "wifi1" in device.device.features:
+        entities.append(
+            DevoloButtonEntity(
+                entry,
+                BUTTON_TYPES[START_WPS],
+                device,
+            )
+        )
+    async_add_entities(entities)
+
+
+class DevoloButtonEntity(DevoloEntity, ButtonEntity):
+    """Representation of a devolo button."""
+
+    def __init__(
+        self,
+        entry: ConfigEntry,
+        description: DevoloButtonEntityDescription,
+        device: Device,
+    ) -> None:
+        """Initialize entity."""
+        self.entity_description: DevoloButtonEntityDescription = description
+        super().__init__(entry, device)
+
+    async def async_press(self) -> None:
+        """Handle the button press."""
+        try:
+            await self.entity_description.press_func(self.device)
+        except DevicePasswordProtected:
+            self.entry.async_start_reauth(self.hass)
```

Comment:
I believe you should raise `ConfigEntryAuthFailed` instead and let HomeAssistant start the reauth flow

Reply:
The other platforms currently do it the same way. Mind if I clean that up all together in a separate MR?

Reply:
You could try raising ConfigEntryAuthFailed here but i think it won't start the reauth flow since nothing is catching it in this flow, but i haven't tried it. My assumption is that you have to do like what this PR does and start it yourself.

But doesn't it also need to raise some failure so that the service call fails explicitly? It looks like it will return and effectively succeed? The test currently exercises that the service call succeeds which seems incorrect.

Reply:
Hi @allenporter , it seems, that there is no other integration yet, that has a button platform and starts a reauth flow. So I looked at e.g. the update platform of shelly, the media player platform of webostv or the switch platform of uptimerobot. None raises an error after starting reauth. My other platforms also don't. However, I spotted raising a HomeAssistantError in the alarm control panel of totalconnect.

I can do it either way, but in case we come to the conclusion to raise an exception, I would like to to that in a follow up PR for all my platforms. What do you think?

Reply:
Here is how you should think about it, ignoring reauth for a moment: If you have a service call and the service call fails, it should raise an exception right? For example, just a line below it raise an exception on `DeviceUnavailable` which is a different type of failure.

Whether or not to fail the call is independent of whether or not you start reauth.

Reply:
Got it. So know I'm raising a HomeAssistantError after starting the reauth flow. I will adapt the other platforms (actually just switch, if I remember correctly) in a separate PR.